Review Slotting As Primitive

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [ ] I have remove the 'EpochIndex' and 'SlotNumber' wrappers (cf comments)
- [ ] I have merged the 'Slotting' module into the 'Primitive' module
- [ ] I have remove the ad-hoc functions for 'SlotId' in favor of a more idiomatic 'Enum' instance

# Comments

<!-- Additional comments or screenshots to attach if any -->

> NOTE: I suggest to review this commit by commit :+1:

This is because we _can't_ actually represent a `SlotNumber` at compile-time with a _satisfactory type_. At a given moment in time, this type has some given bounds. Yet, it can change at any moment and considering it fixed would be a mistake. Plus, we can already encapsulate most of the slotting arithmetic within the `SlotId` type, so we don't _need_ the extra ones for the `EpochIndex` and `SlotNumber` and we can simply avoid manipulating them overall.

Same goes for the ad-hoc functions to increment and decrement slots. In practice, we'll probably never use them as this arithmetic will be done by the nodes and for us, slots are really just some ids / foreign keys on a block. Also, I've define 'Enum' instance despite this instance being _partial_ on the boundaries like for many types in Haskell. This belongs to those gray areas that divide people. I belong to the school of thoughts that believe the type system should help us as much as possible; though, when it truly gets in the way, I find it more reasonable to add a guard and be conscious about the edge case. Here, the only thing that can go wrong is trying to get the successor of the very first slot. Because of this, we have to carry some `Maybe` type everywhere which was, anyway, poorly handled in for the `Nothing` branch in the mock test because in practice, handling the first slot requires extra logic ANYWAY. For the other boundary, even if we were processing blocks at a pace of 1 epoch per hour, it would still take us 38K MILLIONS of year to reach that bound. We've time to see it coming. 

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
Remove 'EpochIndex' and 'SlotNumber' wrapper and make use of 'SlotId' instead when it made sense

This is because we _can't_ actually represent a SlotNumber at compile-time as we would like. At a given moment in time, this type has some given bounds.
Yet, it can change at any moment and considering it fixed would be a mistake. Plus, we can already encapsulate most of the slotting arithmetic within
the 'SlotId' type, so we don't _need_ the extra ones for the EpochIndex and SlotNumber and we can simply avoid manipulating them overall.

KtorZ committed:
merge 'Slotting' module with 'Primitive'

KtorZ committed:
define and use 'Enum' interface for slot ids

Anviking reviewed (commented):
Hm, interesting! This seems like basically the `SlotId` (interface), we'd need when taking `slotsPerEpoch` into account?

KtorZ committed:
Review modules organization and folder architecture for Network Layer

There was a bit of confusion IMO about where the network layer sits
(understandable because there's no clear layout of what the architecture
should / could be anywhere ...).

Also, I took the opportunity to remove the 'Rust' part of the module
names since this is quite irrelevant. We are programming to an interface
(a.k.a the http-bridge API, and the implementation of that interface
shouldn't really matter).

Doing so, I've remove the unnecessary abstraction of the chain producer
Monad. We won't be instantiating multiple network layer at once, so for
now, this adds an extra layer of complexity that isn't needed. In
practice we can just fallback to a plain data type that we instantiate
using one service (e.g. the cardano-http-bridge) or another (e.g. the
Haskell shelley node).

KtorZ committed:
remove now obsolete chain producer modules

KtorZ committed:
move slotting comparison from HttpBridge to Primitive

Anviking reviewed (approved):
Nice commits!

1. `SlotId` — I like it
2. Isn't there a beauty to having Slotting and Primitives separate? If we'd want to use alternate versions of them in tests or something? I guess the rest of the code might view them both as "primitives" though 👍 
3. What is the purpose of the `Enum` instance? Backfilling? Just mocking? This won't work well with changing slotsPerEpoch later? Or will it?

Regardless, lgtm!

Edit, saw:
> I have merged MockNetworkLayer with the RustHttpBridgeSpec, and the RustHttpBridgeClient with RustHttpBridge. Generally speaking, I'd like to avoid us to spread too much with modules, it makes code hard to find in the long run and eventually end up with us duplicating functionality and logic in various places. If there's no good reason to have multiple modules, I'd rather stick as much as we can inside one.

rvl reviewed (approved):
Looks good, thanks.

- You could have added these suggestions as a review comment on the original PR and I would easily have been able to fix it then and saved you some work.
- The arguments against multiple smaller modules are weak.
- SlotId is better without newtype wrappers, but why not type aliases?

rvl committed:
Merge pull request #60 from input-output-hk/KtorZ/review-network-layer

Review Network Layer

rvl commented on src/Cardano/Wallet/Primitive.hs:
```diff
@@ -283,6 +294,84 @@ restrictedTo (UTxO utxo) outs =
+
+-- | Add a number of slots to an (Epoch, LocalSlotIndex) pair, where the number
+-- of slots can be greater than one epoch.
+slotIncr :: Natural -> SlotId -> SlotId
```
Why `Natural`, rather than `Word64`?

KtorZ replied:
I kept the original type-signature defined earlier. I believe that the rational was that `Natural` are unbounded, though in practice, a `Word64` is more than fine indeed.

rvl commented on src/Cardano/Wallet/Primitive.hs:
```diff
@@ -283,6 +294,84 @@ restrictedTo (UTxO utxo) outs =
+-- | Calculates which epochs to fetch, given a number of slots, and the start
+-- point. It takes into account the latest block available, and that the most
+-- recent epoch is never available in a pack file.
+epochRange
```
This function is not really a general-purpose primitive, so doesn't belong here.
The `tipEpoch - 1` bit is because the current epoch is never packed on cardano-http-bridge.

KtorZ replied:
Indeed, I'll move that back onto the HttpBridge :+1:

KtorZ committed:
Move 'epochRange' back to HttpBridge (implementation specific) and use 'Word64' instead of Natural
